homeassistant/components/forked_daapd/media_player.py

```diff
@@ -0,0 +1,673 @@
+                        playback_from_position=saved_queue_position,
+                    )
+                    await self._api.seek(position_ms=saved_song_position)
+                    if saved_state != STATE_PLAYING:
+                        await self.async_media_pause()
+        else:
+            _LOGGER.debug("Media type '%s' not supported.", media_type)
+
+
+class ForkedDaapdUpdater:
+    """Manage updates for the forked-daapd device."""
+
+    def __init__(self, hass, api):
+        """Initialize."""
+        self.hass = hass
+        self._api = api
+        self.websocket_handler = None
+        self._all_output_ids = []
+
+    async def async_init(self):
+        """Perform async portion of class initialization."""
+        server_config = await self._api.get_request("config")
+        websocket_port = server_config.get("websocket_port")
+        if websocket_port:
+            self.websocket_handler = asyncio.create_task(
+                self._api.start_websocket_handler(
+                    server_config["websocket_port"],
+                    WS_NOTIFY_EVENT_TYPES,
+                    self._update,
+                    WEBSOCKET_RECONNECT_TIME,
+                    self._disconnected_callback,
+                )
+            )
+        else:
+            _LOGGER.error("Invalid websocket port.")
+
+    def _disconnected_callback(self):
+        async_dispatcher_send(self.hass, SIGNAL_UPDATE_MASTER, False)
+        async_dispatcher_send(self.hass, SIGNAL_UPDATE_OUTPUTS, [])
+
+    async def _update(self, update_types):
+        """Private update method."""
+
+        def intersect(list1, list2):  # local helper
+            intersection = [i for i in list1 + list2 if i in list1 and i in list2]
+            return bool(intersection)
+
+        _LOGGER.debug("Updating %s", update_types)
+        if "queue" in update_types:  # update queue
+            queue = await self._api.get_request("queue")
+            async_dispatcher_send(self.hass, SIGNAL_UPDATE_QUEUE, queue)
+            await self.hass.async_block_till_done()  # make sure queue done before player for async_play_media
+        # order of below don't matter
+        if intersect(["player", "options", "volume"], update_types):  # update player
+            player = await self._api.get_request("player")
+            async_dispatcher_send(self.hass, SIGNAL_UPDATE_PLAYER, player)
+        if intersect(["outputs", "volume"], update_types):  # update outputs
+            outputs = (await self._api.get_request("outputs"))["outputs"]
+            await self._add_zones(outputs)
+            async_dispatcher_send(self.hass, SIGNAL_UPDATE_OUTPUTS, outputs)
+        if intersect(["database", "update", "config"], update_types):  # not supported
+            _LOGGER.debug(
+                "database/update notifications neither requested nor supported"
+            )
+        await self.hass.async_block_till_done()  # make sure callbacks done before requesting state update
```

Comment:
Should we instead have the entities signal the updater when they are ready, and use an `asyncio.Event` to have the updater wait for the signal to set the event to allow the updater to continue?

Reply:
Sure, I can make the change to use asyncio events.